pallets/order-book/src/lib.rs

```diff
@@ -0,0 +1,541 @@
+			)?;
+			Ok(())
+		}
+
+		#[pallet::call_index(1)]
+		// dummy weight for now
+		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(2, 2).ref_time())]
+		pub fn user_cancel_order(origin: OriginFor<T>, order_id: T::Hash) -> DispatchResult {
+			let account_id = ensure_signed(origin)?;
+			// verify order matches account
+			// UserOrders using Resultquery, if signed account
+			// does not match user for order id, we will get an Err Result
+			Self::cancel_order(order_id)?;
+			Ok(())
+		}
+
+		#[pallet::call_index(2)]
+		// dummy weight for now
+		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(2, 2).ref_time())]
+		pub fn fill_order_full(origin: OriginFor<T>, order_id: T::Hash) -> DispatchResult {
+			let account_id = ensure_signed(origin)?;
+			let order = <Orders<T>>::get(order_id)?;
+			// maybe move to ensure if we don't need these later
+			// might need decimals from currency, but should hopefully be able to use FP
+			// price/amounts from FP balance
+			let asset_out = T::AssetRegistry::metadata(&order.asset_out_id)
+				.ok_or(Error::<T>::InvalidAssetId)?;
+
+			let asset_in =
+				T::AssetRegistry::metadata(&order.asset_in_id).ok_or(Error::<T>::InvalidAssetId)?;
+
+			let sell_amount = order.buy_amount.ensure_mul(order.price)?;
+
+			ensure!(
+				T::TradeableAsset::can_reserve(order.asset_out_id, &account_id, sell_amount),
+				Error::<T>::InsufficientAssetFunds,
+			);
+			T::TradeableAsset::unreserve(
+				order.asset_in_id,
+				&order.placing_account,
+				order.buy_amount,
+			);
+			T::ReserveCurrency::unreserve(&account_id, T::Fees::fee_value(T::OrderFeeKey::get()));
+			T::TradeableAsset::transfer(
+				order.asset_in_id,
+				&account_id,
+				&order.placing_account,
+				order.buy_amount,
+			)?;
+			T::TradeableAsset::transfer(
+				order.asset_out_id,
+				&order.placing_account,
+				&account_id,
+				sell_amount,
+			)?;
+			Self::deposit_event(Event::OrderFulfillment {
+				order_id,
+				placing_account: order.placing_account,
+				fulfilling_account: account_id,
+				partial_fulfillment: true,
+				currency_in: order.asset_in_id,
+				currency_out: order.asset_out_id,
+				fulfillment_amount: order.buy_amount,
+				sell_price_limit: order.price,
+			});
+
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Get all orders for an account
+		/// Provided for frontend to grab orders for an individual account
+		pub fn get_account_orders(
+			account_id: T::AccountId,
+		) -> Result<
+			sp_std::vec::Vec<(
+				T::Hash,
+				Order<T::Hash, T::AccountId, T::AssetCurrencyId, T::ForeignCurrencyBalance>,
+			)>,
+			Error<T>,
+		> {
+			Ok(<UserOrders<T>>::iter_prefix(account_id).collect())
+		}
+
+		/// Get all orders
+		/// Provided for frontend to grab all open orders
+		pub fn get_all_orders() -> Result<
+			sp_std::vec::Vec<(
+				T::Hash,
+				Order<T::Hash, T::AccountId, T::AssetCurrencyId, T::ForeignCurrencyBalance>,
+			)>,
+			Error<T>,
+		> {
+			Ok(<Orders<T>>::iter().collect())
+		}
```

Comment:
You might want to use this for a RPC to let it be queried by the front end easier, right? But can't you just access the storage items `Orders` and `UserOrders` directly instead of using an RPC? I always think if you have complex computations, it's cool to provide a RPC for it, but just querying the storage might not be a big deal for the front end. In addition: If you access the functions inside the runtime, I think you know that `iter...().collect()` is a bad practice and usually leads to overweight of blocks. But I assume the functions will never be called inside the runtime.

Reply:
That's a very good point, as it is only for frontend it won't be called in the runtime, and therefore won't affect blockweights, but definitely a good thing to check! :)

Reply:
Agree with @Chralt98, I would remove these methods to avoid bad usage from the runtime. If they're needed for testing, the tester can call `iter_prefix()` or `iter()` directly.

Forcing to use `iter` or `iter_prefix` the user can see that this implies a lot of read accesses.